### PR TITLE
Add more complete type annotations in polars interpreter

### PR DESCRIPTION
**Pull request by wence-**
## Description

We can check this with:

    pyright --verifytypes cudf_polars --ignoreexternal

Which reports a "type completeness" score of around 94%. This will
improve once pylibcudf gets type stubs.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by wence-** [comment]
No functional change intended.

**Review by jameslamb — APPROVED** [review]
Approving the packaging changes, looks good to me. I'm happy to see us adding a `py.typed` 😁

File: python/cudf_polars/pyproject.toml [hunk 1]
```diff
@@ -62,8 +62,6 @@ target-version = "py39"
 fix = true
 
 [tool.ruff.lint]
-# __init__.py must re-export everything it imports
-ignore-init-module-imports = false
```
Comment (wence-) [inline python/cudf_polars/pyproject.toml:66]: This option is deprecated in ruff

File: python/cudf_polars/cudf_polars/dsl/expr.py [hunk 2]
```diff
@@ -284,9 +284,13 @@ class NamedExpr:
...
     # when evaluating expressions themselves, only when constructing
     # named return values in dataframe (IR) nodes.
-    __slots__ = ("name", "value")
+    __slots__ = ("dtype", "name", "value")
+    value: Expr
+    name: str
+    dtype: plc.DataType
 
     def __init__(self, name: str, value: Expr) -> None:
+        self.dtype = value.dtype
```
Comment (wence-) [inline python/cudf_polars/cudf_polars/dsl/expr.py:293]: I think I never need the dtype so should probably drop this part.

File: python/cudf_polars/cudf_polars/dsl/expr.py [hunk 3]
```diff
@@ -367,6 +372,7 @@ class Col(Expr):
     __slots__ = ("name",)
     _non_child = ("dtype", "name")
     name: str
+    children: tuple[()]
```
Comment (mroeschke) [inline python/cudf_polars/cudf_polars/dsl/expr.py:375]: non-blocking: Does pyright not like just bare `: tuple`?

Comment (mroeschke) [reply python/cudf_polars/cudf_polars/dsl/expr.py:375]: Ah OK I misunderstood the implication here. Thanks

**Review by mroeschke — APPROVED** [review]

**Pull request merged**